### ISSUE TITLE
RuntimeWarning: invalid value when disabling AVX2

### ISSUE THREAD

**Issue by bmerry** [numba/numba#6819]
## Reporting a bug

- [x] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

Test code:
```python
#!/usr/bin/env python3

import os

os.environ['NUMBA_CPU_FEATURES'] = '-avx2'

import numpy as np
import numba

@numba.guvectorize(['c8[:], c8[:]'], '(a)->(a)', nopython=True)
def _stefcal_gufunc(init_gain, g):
    print(np.complex64(1) / init_gain[0])

_stefcal_gufunc(np.ones(1, np.complex64))
```

Output
```
(1+0j)
/home/bmerry/work/sdp/env3/lib/python3.8/site-packages/numba/np/ufunc/gufunc.py:123: RuntimeWarning: invalid value encountered in _stefcal_gufunc
  return self.ufunc(*args, **kwargs)
```

I expected there to be no RuntimeWarning since it's just computing 1/1. Removing the NUMBA_CPU_FEATURES line (on a machine with avx2) makes the warning go away.

This appears to be a regression from 0.51.2 to 0.52.

- Python 3.8
- numpy 1.20.1
- numba 0.53.0
- llvmlite 0.36.0
- Everything installed with pip

**Comment by bmerry** [comment]
@stuartarchibald this is the bug I mentioned on Gitter.

**Comment by esc** [comment]
<img width="499" alt="Screen Shot 2021-03-12 at 20 54 38" src="https://user-images.githubusercontent.com/41563/110991643-2f207280-8375-11eb-8603-159a1d7c46c3.png">

Classifying as a bug.

**Comment by esc** [comment]
@bmerry thanks for submitting this to the Numba issue tracking. Judging from @stuartarchibald 's comment on Gitter this is a bug.

**Comment by bmerry** [comment]
What are the implications: is it just a spurious warning, or will it lead to incorrect results?

**Comment by stuartarchibald** [comment]
This also triggers it:
```python
import numba
import numpy as np

NAN = float('nan')
DEAD = np.float64(-1.1935504820988231e+148) # -2401018187971961171 in LC pool

@numba.njit
def complex_div(a, b):
    areal = a.real
    aimag = a.imag
    breal = b.real
    bimag = b.imag

    if abs(breal) > DEAD: # DEAD as machine code identifier
        ratio = (bimag / breal)
        denom = breal + (bimag * ratio)
        return complex((areal + aimag * ratio) / denom,
                       (aimag + areal * ratio) / denom)
    else:
        return complex(0, 0)

@numba.guvectorize(['c8[:], c8[:]'], '(a)->(a)', nopython=True)
def foo(init_gain, g):
    g[0] = complex_div(np.complex64(1) , init_gain[0])

init_gain = np.full((1,), 3., dtype=np.complex64)
g = np.zeros_like(init_gain)

foo(init_gain, g)
```
I don't _think_ the answer is wrong (needs more investigation). I suspect it's just that some optimisation has ended up with a sequence being run that cause a FPE which NumPy picks up and doing things like changing the instruction set changes the optimisation patterns which stops that particular sequence occurring and executing.

**Comment by ludwigschwardt** [comment]
Some more random data points...

### TL;DR
- It looks like @bmerry's initial test code involving the `-avx2` flag now passes on numba 0.57.0 and 0.57.1rc1, verified on both numpy 1.21.0 and 1.24.3.
- I could never get Stuart's snippet above to fail, unless it directly followed after Bruce's code and therefore also gets the `-avx2` flag. Was that the intention, @stuartarchibald?

This gives me some hope that we can finally move on from numba 0.51.2 😅 

### Details

I refer to Bruce's initial test code as A and to Stuart's test code in the above comment as B. I only refer to tests I actually ran.

The following fails:
- numba 0.53.0, llvmlite 0.36.0, numpy 1.23.5, A
- numba 0.54.0, llvmlite 0.37.0, numpy 1.20.3, A
- numba 0.55.0, llvmlite 0.38.1, numpy 1.20.3, A
- numba 0.55.2, llvmlite 0.38.1, numpy 1.20.3, A
- numba 0.56.0, llvmlite 0.39.1, numpy 1.20.3, A
- numba 0.56.2, llvmlite 0.39.1, numpy 1.20.3, A
- numba 0.56.3, llvmlite 0.39.1, numpy 1.20.3, A
- numba 0.56.3, llvmlite 0.39.1, numpy 1.23.5, A
- numba 0.56.4, llvmlite 0.39.1, numpy 1.23.5, A

The following passes:
- numba 0.55.2, llvmlite 0.38.1, numpy 1.20.3, B
- numba 0.56.0, llvmlite 0.39.1, numpy 1.20.3, B
- numba 0.56.2, llvmlite 0.39.1, numpy 1.20.3, B
- numba 0.56.3, llvmlite 0.39.1, numpy 1.20.3, B
- numba 0.56.4, llvmlite 0.39.1, numpy 1.20.3, B
- numba 0.56.4, llvmlite 0.39.1, numpy 1.23.5, B
- numba 0.57.0, llvmlite 0.40.1rc1, numpy 1.21.0, A + B
- numba 0.57.0, llvmlite 0.40.1rc1, numpy 1.24.3, A + B
- numba 0.57.1rc1, llvmlite 0.40.1rc1, numpy 1.21.0, A + B
- numba 0.57.1rc1, llvmlite 0.40.1rc1, numpy 1.24.3, A + B
